gateway-shell/src/main/java/org/apache/knox/gateway/shell/KnoxSh.java

```diff
@@ -155,6 +170,81 @@ public boolean validate() {
+    private static final String CLIENT_TRUST_STORE_FILE_NAME = "gateway-client-trust.jks";
+
+    @Override
+    public void execute() throws Exception {
+      final X509Certificate gatewayServerPublicCert = fetchPublicCertFromGatewayServer();
+      if (gatewayServerPublicCert != null) {
+        final File trustStoreFile = new File(System.getProperty("user.home"), CLIENT_TRUST_STORE_FILE_NAME);
+        X509CertificateUtil.writeCertificateToJks(gatewayServerPublicCert, trustStoreFile);
+        out.println("Gateway server's certificate is exported into " + trustStoreFile.getAbsolutePath());
+      } else {
+        out.println("Could not obtain server certificate chain");
+      }
+    }
+
+    private X509Certificate fetchPublicCertFromGatewayServer() throws Exception {
+      final TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+      trustManagerFactory.init((KeyStore) null);
+      final X509TrustManager defaultTrustManager = (X509TrustManager) trustManagerFactory.getTrustManagers()[0];
+      final CertificateChainAwareTrustManager trustManagerWithCertificateChain = new CertificateChainAwareTrustManager(defaultTrustManager);
+      final SSLContext sslContext = SSLContext.getInstance("TLS");
+      sslContext.init(null, new TrustManager[] { trustManagerWithCertificateChain }, null);
+
+      final URI uri = URI.create(gateway);
+      out.println("Opening connection to " + uri.getHost() + ":" + uri.getPort() + "...");
+      final SSLSocket socket = (SSLSocket) sslContext.getSocketFactory().createSocket(uri.getHost(), uri.getPort());
+      socket.setSoTimeout(10000);
+      try {
+        out.println("Starting SSL handshake...");
+        socket.startHandshake();
+        socket.close();
+        out.println();
+        out.println("No errors, certificate is already trusted");
+      } catch (SSLException e) {
+        // NOP; this is expected in case the gateway server's certificate is not in the trust store the JVM uses
+        out.println("SSL exception; found non-trusted certificate");
+      }
+
+      return trustManagerWithCertificateChain.certificateChain == null ? null : trustManagerWithCertificateChain.certificateChain[0];
+    }
+
+    @Override
+    public String getUsage() {
+      return USAGE + ":\n\n" + DESC;
+    }
+
+    private class CertificateChainAwareTrustManager implements X509TrustManager {
+      private final X509TrustManager defaultTrustManager;
+      private X509Certificate[] certificateChain;
+
+      CertificateChainAwareTrustManager(X509TrustManager tm) {
+        this.defaultTrustManager = tm;
+      }
+
+      @Override
+      public X509Certificate[] getAcceptedIssuers() {
+        throw new UnsupportedOperationException();
+      }
+
+      @Override
+      public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        throw new UnsupportedOperationException();
```

Comment:
If Knox is setup to require client auth? Does this break?

Reply:
Yes; I'll fix it.